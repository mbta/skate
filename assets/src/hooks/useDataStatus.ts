import { Channel, Socket } from "phoenix"
import { useState, useEffect } from "react"
import { reload } from "../models/browser"

export type DataStatus = "good" | "outage"

const topic = "data_status"

const useDataStatus = (socket: Socket | undefined) => {
  const [state, setState] = useState<DataStatus>("good")

  useEffect(() => {
    if (socket !== undefined) {
      const channel: Channel = socket.channel(topic)
      channel.on("data_status", ({ data: status }) => {
        setState(status)
      })
      channel
        .join()
        .receive("ok", ({ data: status }) => {
          setState(status)
        })
        // tslint:disable-next-line: no-console
        .receive("error", ({ reason }) => console.error("join failed", reason))
        .receive("timeout", () => {
          reload(true)
        })
    }
  }, [socket])
  return state
}

export default useDataStatus
