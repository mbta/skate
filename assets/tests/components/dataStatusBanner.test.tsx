import React from "react"
import renderer from "react-test-renderer"
import DataStatusBanner from "../../src/components/dataStatusBanner"
import useDataStatus from "../../src/hooks/useDataStatus"

jest.mock("../../src/hooks/useDataStatus", () => ({
  __esModule: true,
  default: jest.fn(),
}))

describe("DataStatusBanner", () => {
  test("doesn't show anything when status is good", () => {
    ;(useDataStatus as jest.Mock).mockImplementationOnce(() => "good")
    const tree = renderer.create(<DataStatusBanner />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test("shows outage banner when there's an outage", () => {
    ;(useDataStatus as jest.Mock).mockImplementationOnce(() => "outage")
    const tree = renderer.create(<DataStatusBanner />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
