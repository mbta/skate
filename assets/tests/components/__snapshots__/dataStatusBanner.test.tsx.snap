// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataStatusBanner doesn't show anything when status is good 1`] = `null`;

exports[`DataStatusBanner shows outage banner when there's an outage 1`] = `
<div
  className="m-data-status-banner m-data-status-banner--outage"
>
  <div
    className="m-data-status-banner__heading"
  >
    Ongoing MBTA Data Outage
  </div>
  <div
    className="m-data-status-banner__content"
  >
    Vehicle information may be missing or inaccurate. Thank you for your patience as we work to fix this issue.
  </div>
</div>
`;
